modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/programs/Download.scala

```diff
@@ -1,30 +1,62 @@
 package org.tessellation.sdk.infrastructure.snapshot.programs
 
+import cats.Monad
 import cats.effect.Async
 import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import scala.concurrent.duration.{Duration, DurationInt}
 
 import org.tessellation.schema.node.NodeState
+import org.tessellation.schema.peer.Peer
 import org.tessellation.sdk.domain.node.NodeStorage
 import org.tessellation.sdk.infrastructure.snapshot.SnapshotConsensus
 
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
 object Download {
 
   def make[F[_]: Async](
     nodeStorage: NodeStorage[F],
-    consensus: SnapshotConsensus[F, _, _, _, _, _]
+    consensus: SnapshotConsensus[F, _, _, _, _, _],
+    getPeers: () => F[Set[Peer]]
   ): Download[F] =
     new Download[F](
       nodeStorage,
-      consensus
+      consensus,
+      getPeers
     ) {}
 }
 
 sealed abstract class Download[F[_]: Async] private (
   nodeStorage: NodeStorage[F],
-  consensus: SnapshotConsensus[F, _, _, _, _, _]
+  consensus: SnapshotConsensus[F, _, _, _, _, _],
+  getPeers: () => F[Set[Peer]]
 ) {
 
+  val logger = Slf4jLogger.getLoggerFromClass[F](Download.getClass)
+
+  private val timeout: Duration = 15.seconds
+  private val minKnownPeers = 50
+  private val delayBetweenPeerCountChecks = 1.second
+
+  private def waitUntilMinPeersFound: F[Unit] = {
+    val notEnoughPeers = getPeers().map(_.size < minKnownPeers)
+    val waitBetweenPeerCountChecks = Async[F].sleep(delayBetweenPeerCountChecks)
+    val peerCountTask = Monad[F].whileM_(notEnoughPeers)(waitBetweenPeerCountChecks) >>
+      logger.debug(s"At least ${minKnownPeers} peers have been discovered.")
+    val fallbackTask = Async[F].unit >>
+      logger.debug(s"${timeout} seconds have passed and ${minKnownPeers} peers have not been found.")
```

Comment:
logger.debug is already of type `F[Unit]` so you don't need to specify it as `Async[F].unit >> logger`